src/tools/rbd/action/Export.cc

```diff
@@ -116,34 +117,92 @@ static int do_export(librbd::Image& image, const char *path, bool no_progress)
   }
 
   utils::ProgressContext pc("Exporting image", no_progress);
-
   SimpleThrottle throttle(max_concurrent_ops, false);
   uint64_t period = image.get_stripe_count() * (1ull << info.order);
+
+  size_t file_size = 0;
+  if (with_snaps) {
+    // header
+    bufferlist bl;
+    bl.append(utils::RBD_IMAGE_BANNER);
+
+    uint64_t size = info.size;
+    ::encode(size, bl);
+
+    r = bl.write_fd(fd);
+    if (r < 0) {
+      goto out;
+    }
+    file_size += bl.length();
+    lseek(fd, file_size, SEEK_SET);
+  }
+
   for (uint64_t offset = 0; offset < info.size; offset += period) {
     if (throttle.pending_error()) {
       break;
     }
 
     uint64_t length = min(period, info.size - offset);
-    C_Export *ctx = new C_Export(throttle, image, offset, length, fd);
+    C_Export *ctx = new C_Export(throttle, image, file_size + offset, offset, length, fd);
     ctx->send();
 
-    pc.update_progress(offset, info.size);
+    pc.update_progress(offset, info.size * 2);
   }
 
+  file_size += info.size;
   r = throttle.wait_for_ret();
   if (!to_stdout) {
     if (r >= 0) {
-      r = ftruncate(fd, info.size);
+      r = ftruncate(fd, file_size);
+      r = lseek(fd, file_size, SEEK_SET);
     }
-    close(fd);
   }
 
-  if (r < 0) {
+  if (with_snaps) {
```

Comment:
But then we can't import a image without snaps in this case.
```
rbd export --with-snaps test test.export
rbd import test.import test.export
```
I mean, sometimes, we can get a exported image with snapshots, but we only want to import a full image from it.